cluster/src/main/scala/org/apache/spark/util/DynamicURLClassLoader.scala

```diff
@@ -0,0 +1,156 @@
+    parentClassLoader: ClassLoader, parentFirst: Boolean, overwriteFiles: Boolean = false)
+    extends MutableURLClassLoader(Array[URL](), parentClassLoader) with  Logging {
+  final private val JOB_SERVER_TEXT = "_SNAPPY_JOB_SERVER_JAR_"
+
+  private val urlList = new util.TreeMap[String, URLClassLoader]()
+
+  // this is executor class Path which is not going to change
+  private val parent = {
+    if (parentFirst) new MutableURLClassLoader(urls, parentClassLoader)
+    else new ChildFirstURLClassLoader(urls, parentClassLoader)
+  }
+
+  override def addURL(url: URL): Unit = {
+    // workaround to identify the files loaded by the job server as
+    // it always creates a new file with timestamp.
+    val key = if (url.getFile.lastIndexOf(JOB_SERVER_TEXT) > 0) {
+      url.getFile.substring(0, url.getFile.lastIndexOf(JOB_SERVER_TEXT))
+    } else url.getFile
+
+    val loader = getClassLoader(url)
+    def oldURL = if (urlList.containsKey(key)) urlList.get(key).getURLs.head.getPath else ""
+
+    if (!oldURL.isEmpty && !oldURL.equals(url.getPath)) {
+      // try to remove the old file. it is  okay if not successful
+      try {
+        logInfo(" Deleting the Loader with key:" + key + " with File: " + oldURL )
+        new File(oldURL).delete()
+      } catch {
+        case _: Throwable => {
+          // do nothing.
+        }
+      }
+    }
+
+    logInfo(" Adding new Loader with key:" + key + " with File: " + url.getPath )
+    urlList.put(key, loader)
+  }
+
+  private def getClassLoader(url: URL): MutableURLClassLoader = {
+    if (parentFirst) new MutableURLClassLoader(Array(url), parent)
+    else new ChildFirstURLClassLoader(Array(url), parent)
+  }
+
+  override def getURLs(): Array[URL] = {
+    urlList.asScala.flatMap(_._2.getURLs).toArray ++ parent.getURLs
+  }
+
+  override def getResources(name: String): util.Enumeration[URL] = {
+    val childResources =
+      urlList.asScala.map(tuple => {
+        tuple._2.getResources(name).asScala
+            .filter(!parent.getResources(name).asScala.contains(_)).asJavaEnumeration
+      }).toArray
+
+    val parentResources = Array(parent.getResources(name).asInstanceOf[util.Enumeration[_]])
+
+    new CompoundEnumeration(childResources ++ parentResources)
+  }
+
+  @throws[ClassNotFoundException]
+  override protected def loadClass(className: String, resolve: Boolean): Class[_] = {
+    var c: Class[_] = findLoadedClass(className)
+
+    if (c != null) {
+      if (resolve) resolveClass(c)
+      return c
+    }
+
+    if (parentFirst) {
+      c = loadFromParent(parent, className, false)
+          .getOrElse(loadFromJars(className, false)
+              .getOrElse(loadFromStore(className, true).get))
+    } else {
+      c = loadFromJars(className, false)
+          .getOrElse(loadFromStore(className, false)
+              .getOrElse(loadFromParent(parent, className, true).get))
+    }
+
+    if (resolve) {
+      resolveClass(c)
+    }
+
+    c
+  }
```

Comment:
This seems to be using Spark loader first, then fallback to store loader. I think there will be less confusion if we use only ours (so job-server can do an install-jar for a new provided jar). Lets discuss.

Secondly why do we need to do explicit loadFromJars too? Will loadFromParent not suffice?